core/common/lib/policy-engine-lib/src/main/java/org/eclipse/edc/policy/engine/validation/PolicyValidator.java

```diff
@@ -0,0 +1,264 @@
+import org.eclipse.edc.policy.model.Constraint;
+import org.eclipse.edc.policy.model.Duty;
+import org.eclipse.edc.policy.model.LiteralExpression;
+import org.eclipse.edc.policy.model.MultiplicityConstraint;
+import org.eclipse.edc.policy.model.Operator;
+import org.eclipse.edc.policy.model.OrConstraint;
+import org.eclipse.edc.policy.model.Permission;
+import org.eclipse.edc.policy.model.Policy;
+import org.eclipse.edc.policy.model.Prohibition;
+import org.eclipse.edc.policy.model.Rule;
+import org.eclipse.edc.policy.model.XoneConstraint;
+import org.eclipse.edc.spi.result.Result;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Stack;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/**
+ * Validate a policy.
+ * <p>
+ * The policy validator is used to validate policies against a set of configured rule bindings, {@link AtomicConstraintFunction} and {@link DynamicAtomicConstraintFunction}
+ * <p>
+ * The validation will fail under the following conditions:
+ *
+ * <ul>
+ *     <li>If a rule action is not bound to a scope.</li>
+ *     <li>If an {@link AtomicConstraint}'s left-operand is not bound to a scope.
+ *     <li>If an {@link AtomicConstraint}'s left-operand is not bound to a function.
+ * </ul>
+ */
+public class PolicyValidator implements Policy.Visitor<Result<Void>>, Rule.Visitor<Result<Void>>, Constraint.Visitor<Result<Void>> {
+
+    private final Stack<Rule> ruleContext = new Stack<>();
+
+    // Key -> Functions
+    private final Map<String, List<ConstraintFunctionEntry<Rule>>> constraintFunctions = new TreeMap<>();
+    private final List<DynamicAtomicConstraintFunctionEntry<Rule>> dynamicConstraintFunctions = new ArrayList<>();
+    private RuleValidator ruleValidator;
+
+    public Result<Void> validate(Policy policy) {
+        return policy.accept(this);
+    }
+
+    @Override
+    public Result<Void> visitAndConstraint(AndConstraint constraint) {
+        return validateMultiplicityConstraint(constraint);
+    }
+
+    @Override
+    public Result<Void> visitOrConstraint(OrConstraint constraint) {
+        return validateMultiplicityConstraint(constraint);
+    }
+
+    @Override
+    public Result<Void> visitXoneConstraint(XoneConstraint constraint) {
+        return validateMultiplicityConstraint(constraint);
+    }
+
+    @Override
+    public Result<Void> visitAtomicConstraint(AtomicConstraint constraint) {
+        var currentRule = currentRule();
+        var leftValue = constraint.getLeftExpression().accept(s -> s.getValue().toString());
+        var rightValue = constraint.getRightExpression().accept(LiteralExpression::getValue);
+
+        return validateLeftExpression(currentRule, leftValue)
+                .merge(validateConstraint(leftValue, constraint.getOperator(), rightValue, currentRule));
+    }
+
+    @Override
+    public Result<Void> visitPolicy(Policy policy) {
```

Comment:
nit: this could be written in a more succinct way like:
```
    @Override
    public Result<Void> visitPolicy(Policy policy) {
        return Stream.of(policy.getPermissions(), policy.getProhibitions(), policy.getObligations())
                .flatMap(Collection::stream)
                .map(rule -> rule.accept(this))
                .reduce(Result.success(), Result::merge);
    }
```